### PR TITLE
Copy an application to another candidate

### PR DESCRIPTION
**Pull request by tomas-stefano**
## Context

This is to solve the incident and to move one application form from one candidate account to another.

## Changes proposed in this pull request

Make duplication application service to move one application form to another (to be used only on exceptional circumstances)

## Gotchas

When we move one application form to another candidate, we need to make sure the previous application form is set correctly.

This graph shows the case and where we should link to the previous application form:

Screenshot 2024-03-11 at 10 54 48

The graph above is assuming that we called:

```ruby
  DuplicateApplication.new(
    fake_application_form,
    candidate_id: real_candidate.id,
  ).duplicate
```

### CODE REVIEW

**Review by dcyoung-dev — APPROVED** [review]
This looks good and will solve our migration troubles 👌 

We should create a linked card to remove this work once we're done with it.

**Comment by dcyoung-dev** [comment]
This code will be removed as part of tidy up https://trello.com/c/J4BQoZ4T

**Comment by inulty-dfe** [comment]
Does this code account for when a candidate has an application form in the current cycle?

**Comment by elceebee** [comment]
I don't want to old things up, and happy to be convinced otherwise, but I feel like, rather than adding a purpose to the existing service, should we 
- create a new service specifically for the incident
- copy over all the relevant logic from the existing service
- add the logic captured in the slack thread about what needs to be done in the rails console. (eg, creating the new account, conditionally copying over references, candidate choices, etc)

That way, the end product requires us just to call one command on the rails console per application. something like `CopyIncidentApplicationToNewAccount.call(application_form, new_account_email: email)`

**Comment by tomas-stefano** [comment]
> Does this code account for when a candidate has an application form in the current cycle?

Yes, it does. One of the contexts tested.

**Comment by tomas-stefano** [comment]
> I don't want to hold things up, and happy to be convinced otherwise, but rather than adding a purpose to the existing service, should we
> 
> * create a new service specifically for the incident
> * copy over all the relevant logic from the existing service
> * add the logic captured in the slack thread about what needs to be done in the rails console. (eg, creating the new account, conditionally copying over references, candidate choices, etc)
> 
> That way, we call one command on the rails console per application. something like `CopyIncidentApplicationToNewAccount.call(application_form, new_account_email: email)`

I like the name! Yes, we didn't opt-in for this path due to many possible states application choices could be in an application form. But I can have a go for the ones we need to run that are up to awaiting decision. I will give it a go!

**Comment by dcyoung-dev** [comment]
> Does this code account for when a candidate has an application form in the current cycle?

Yeah we copy the application and replace the application in the current cycle - we probably don't have to worry as the Applications that we're running this on are known and we'll check each one before running this 🤞 

https://github.com/DFE-Digital/apply-for-teacher-training/pull/9186/files#diff-22aafc8535bd72d3c19ad722f739216432316b97557f94d1e937098c7c89b715R156

File: app/services/copy_incident_application_to_new_account.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
+class CopyIncidentApplicationToNewAccount
+  attr_reader :original_application_form, :candidate_email_address
+
+  def initialize(original_application_form:, candidate_email_address:)
+    @original_application_form = original_application_form
+    @candidate_email_address = candidate_email_address
+  end
+
+  def call
```
Comment (dcyoung-dev) [inline app/services/copy_incident_application_to_new_account.rb:9]:
```suggestion
  def call!
```

File: spec/services/copy_incident_application_to_new_account_spec.rb [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+      expect(duplicate_application_form).to be_becoming_a_teacher_completed
+      expect(duplicate_application_form.candidate_id).to be(another_candidate.id)
+      expect(duplicate_application_form.recruitment_cycle_year).to be(RecruitmentCycle.current_year)
+      expect(another_candidate.current_application.previous_application_form).to be_nil
+    end
+  end
+
+  context 'when candidate does have an application form in previous cycle' do
+    it 'assigns the original application to the candidate in current cycle' do
```
Comment (dcyoung-dev) [inline spec/services/copy_incident_application_to_new_account_spec.rb:37]:
```suggestion
    it 'copies the original application to the new candidate in current recruitment cycle with a link to their previous application' do
```

File: app/services/copy_incident_application_to_new_account.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+      new_application_form.application_choices.new.configure_initial_course_choice!(course_option)
+    end
+
+    # Submit application choices - only from awaiting provider decisions - from
+    # original application form
+    new_application_form.application_choices.each do |application_choice|
+      original_application_choice = original_application_form.application_choices.find_by(course_option: application_choice.course_option)
+      next unless original_application_choice.awaiting_provider_decision?
+
+      CandidateInterface::ContinuousApplications::SubmitApplicationChoice.new(application_choice).call
```
Comment (inulty-dfe) [inline app/services/copy_incident_application_to_new_account.rb:42]: What if the courses are closed now, or otherwise invalid?

Comment (tomas-stefano) [reply app/services/copy_incident_application_to_new_account.rb:42]: Good question. I will use the ApplicationChoiceSubmission service to check everything is good to submit.

**Review by dcyoung-dev — COMMENTED** [review]
Some minor suggestions

File: app/services/copy_incident_application_to_new_account.rb (hunk 3 above)
Comment (tomas-stefano) [reply app/services/copy_incident_application_to_new_account.rb]: Done!

**Comment by inulty-dfe** [comment]
Similar to your suggestion on the eviction PR, it might be good to perform some logging in this service.

File: app/services/copy_incident_application_to_new_account.rb [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    @original_application_form = original_application_form
+    @candidate_email_address = candidate_email_address
+  end
+
+  def call!
+    # Create a new Candidate
+    # this will not create a duplicate if they already exist
+    log('Creating a new candidate if they do not exist')
+    CandidateInterface::SignUpForm.new(email_address: candidate_email_address).save
+    candidate = Candidate.find_by!(email_address: candidate_email_address)
```
Comment (dcyoung-dev) [inline app/services/copy_incident_application_to_new_account.rb:14]: The email address has to be in all lowercase here 🤦

Comment (tomas-stefano) [reply app/services/copy_incident_application_to_new_account.rb:14]: Done

**Pull request merged**